Reducing the amount of time ConsoleLogger locks

DavidObando opened the pull request:
The code has been improved by:
 - Pre-computing the state before locking
 - Reducing the amount of allocations of color objects
 - Reducing the amount of calls to Console.Write to the minimum, while still supporting colors.

Tests have been adjusted after the changes, and a couple new tests were added. This PR is intended to be merged into the 'dev' branch.

/cc @rynowak @sivagms @anpete @Eilon @SajayAntony @mnltejaswini @muratg @DamianEdwards @danroth27 @divega

dnfclas commented:
Hi __@DavidObando__, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by .NET Foundation and real humans are currently evaluating your PR.</p>

TTYL, DNFBOT;

rynowak commented:
ConsoleColors is a struct.

BrennanConroy commented on src/Microsoft.Extensions.Logging.Console/ConsoleLogger.cs:
```diff
@@ -15,10 +15,20 @@ public class ConsoleLogger : ILogger
 
+        // do not change user's background color except for Critical
+        private readonly ConsoleColors CriticalConsoleColors = new ConsoleColors(ConsoleColor.White, ConsoleColor.Red);
+        private readonly ConsoleColors ErrorConsoleColors = new ConsoleColors(ConsoleColor.Red, ConsoleColor.Black);
```
ConsoleColor.Black != user's background color

BrennanConroy replied:
Oh, I think the comment is outdated, we made a decision to set black as background for everything to make sure the foreground color was visible

DavidObando replied:
@BrennanConroy 

> ConsoleColor.Black != user's background color

That's what was there before. I just made them instance members instead of producing a new ConsoleColors instance each time it was needed. (see [here](https://github.com/aspnet/Logging/blob/3d853c72eb9fb7e6182840d5c7811769dd431c62/src/Microsoft.Extensions.Logging.Console/ConsoleLogger.cs#L239))

> Oh, I think the comment is outdated, we made a decision to set black as background for everything to make sure the foreground color was visible

Idem. This comment was simply moved with the colors. (see [here](https://github.com/aspnet/Logging/blob/3d853c72eb9fb7e6182840d5c7811769dd431c62/src/Microsoft.Extensions.Logging.Console/ConsoleLogger.cs#L233))

BrennanConroy replied:
Yes I realize that, but the comment should have been updated back when I made some changes, hence the comment is outdated

DavidObando replied:
@BrennanConroy so does that mean that `CriticalConsoleColors = new ConsoleColors(ConsoleColor.White, ConsoleColor.Red);` is wrong? Should it be white on black instead of white on red?

BrennanConroy replied:
They're all correct. Should change the comment to something like "We must explicitly set the background color if we are setting the foreground color otherwise just setting one can look bad on the users console"

Eilon commented on src/Microsoft.Extensions.Logging.Console/ConsoleLogger.cs:
```diff
@@ -97,101 +108,116 @@ public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Except
-            var loglevelString = GetLogLevelString(logLevel);
+            var fullMessageBuilder = new StringBuilder();
+
+            ConsoleColors logLevelColors = DefaultConsoleColors;
```
`var`, `var`, `var`...

Eilon commented on src/Microsoft.Extensions.Logging.Console/ConsoleLogger.cs:
```diff
@@ -97,101 +108,116 @@ public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Except
             //       Request received
-
-            lock (_lock)
+            if(!string.IsNullOrEmpty(message))
```
Formatting

Eilon commented on src/Microsoft.Extensions.Logging.Console/ConsoleLogger.cs:
```diff
@@ -97,101 +108,116 @@ public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Except
-                    DefaultConsoleColor,
-                    _loglevelPadding + logName + $"[{eventId}]",
-                    newLine: true);
+                if (loglevelUsesDefaultColors)
```
Why is this `if` here? It wasn't there before, was it?

BrennanConroy replied:
I understand it now, however it seems you don't need an if and can just assume the logLevel doesn't use default colors?

DavidObando replied:
You're right. I was trying to fold the log level into the full message when it used default colors, but that case appears to not be true ever under the current implementation. I'll change this to always print the log level (special colors) separately from the full message (default colors).

Eilon commented on src/Microsoft.Extensions.Logging.Console/ConsoleLogger.cs:
```diff
@@ -97,101 +108,116 @@ public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Except
+            var fullMessageBuilder = new StringBuilder();
+
+            ConsoleColors logLevelColors = DefaultConsoleColors;
+            string loglevelString = string.Empty;
```
The var names should be `logLevelXyz` (`Level` starts with capital `L`)

rynowak commented:
Have any profiling/allocation data for this change?

It's definitely a good change to move all the logic outside of the lock, but there's a tradeoff being made here that I don't think we want to make with the use of the `StringBuilder`. That's going to allocate roughly 2x the amount of memory needed to represent the message (`char[]` in the `StringBuilder` + the final string for the whole message). Most of the things we need to write are already strings.

Breaking this up into a series of individual writes can avoid that, at the cost of taking the underlying console lock multiple times (which should be uncontended).

Eilon commented:
I think one must-have is that we lock the console for an entire log message. Otherwise it will interleave multiple log calls and inter-mix them. While it would be "thread-safe," it would bring back the exact problem we must avoid.

rynowak commented:
>I think one must-have is that we lock the console for an entire log message.

Yep, I agree.

My comments refer to the locking done inside the console APIs. Every call to `Console.Write` does a [volatile read](https://github.com/dotnet/corefx/blob/master/src/System.Console/src/System/Console.cs#L133) and then calls into a [synchronized textwriter](https://github.com/dotnet/corefx/blob/master/src/System.Console/src/System/Console.cs#L433).

The tradeoff is whether or not to build as large of chunks of text as possible (this PR), or to do smaller writes to the console based on the pieces we already have (my suggestion). This would all be done while holding 'our' lock regardless. We'd have to measure to know, but this PR as is will allocate a significant amount of memory that we're not allocating today.

DavidObando commented:
:up::date:

I went ahead and did an allocation analysis for this change and confirmed that the pressure added by using the StringBuilder is enough to make an impact, so I went ahead and removed it in favor of having multiple calls to Console.Write, as @rynowak suggested. I also ran a short haul stress test on it via the MusicStore end-to-end scenario, and the results are satisfactory.

I sent an update to this PR addressing the all feedback obtained so far, let me know if you have additional observations.

BrennanConroy commented on src/Microsoft.Extensions.Logging.Console/ConsoleLogger.cs:
```diff
@@ -97,99 +97,104 @@ public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Except
+                    }
+
+                    // use default colors from here on
+                    if (!string.IsNullOrEmpty(logLevelString))
```
Shouldn't this check be `logIdentifier`? And if yes, then it shouldn't need a check since it always has something.

DavidObando replied:
Thanks. Yes, it should be `logIdentifier`, but it may be empty when the Log method is invoked with an Exception object and a null message string.

BrennanConroy replied:
Oh sorry, it only has something if the message had something. But still I think the check is wrong?

DavidObando replied:
It is wrong. Will fix.

BrennanConroy commented on src/Microsoft.Extensions.Logging.Console/ConsoleLogger.cs:
```diff
@@ -97,99 +97,104 @@ public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Except
+                printException = true;
+            }
+
+            if (printMessage || printException)
```
We could just use 1 bool "printLog"

DavidObando replied:
Agreed.

BrennanConroy commented:
Looks good

kichalla commented on src/Microsoft.Extensions.Logging.Console/ConsoleLogger.cs:
```diff
@@ -97,99 +97,104 @@ public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Except
                 if (IncludeScopes)
                 {
-                    var scopeInformation = GetScopeInformation();
+                    scopeInformation = GetScopeInformation();
```
Since this call is now moved outside the lock, wondering if its going to mess up the data. There could be multiple threads each with its own scope tree, but while one of threads gains the lock and is trying to write to the console, another thread could modify the scope tree information giving misleading results. I think this should stay within the lock as before?

rynowak replied:
`scopeInformation` is going to access TLS/ExecutionContext. It's the opposite of the usual problem of threading. 

> another thread could modify the scope tree information giving 

This won't happen. The scope information is not accessible to another thread.

kichalla replied:
Cool. Thanks.

kichalla commented:
:watch:

DavidObando commented:
:up: :date: Let me know if this looks good. Thanks!

BrennanConroy commented on src/Microsoft.Extensions.Logging.Console/ConsoleLogger.cs:
```diff
@@ -97,99 +97,103 @@ public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Except
+                    // use default colors from here on
+                    if (!string.IsNullOrEmpty(logIdentifier))
+                    {
+                        Console.WriteLine(
```
You **_could_** put this in the previous if check since if `logLevelString` is set then `logIdentifier` is set too, but that might be a bit too much :)

BrennanConroy commented:
Wait a few hours for others to have time to review, otherwise :shipit:

Eilon commented:
:shipit:
